cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,540 @@
+
+        # translate exit status
+        if my_status == 0: # found unsatisfiability
+            if self.objective_given and self.xct_solver.hasSolution():
+                self.cpm_status.exitstatus = ExitStatus.OPTIMAL
+            else:
+                self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        elif my_status == 1: # found solution, but not optimality proven
+            assert self.xct_solver.hasSolution()
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        elif my_status == 2: # found inconsistency over assumptions
+            self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        elif my_status == 3: # found timeout
+            self.cpm_status.exitstatus = ExitStatus.UNKNOWN
+            return 0
+        else:
+            raise NotImplementedError(my_status)  # a new status type was introduced, please report on github
+
+        return self._fillObjAndVars()
+
+    def solveAll(self, display=None, time_limit=None, solution_limit=None, call_from_model=False, **kwargs):
+        """
+            Compute all solutions and optionally display the solutions.
+
+            This is the generic implementation, solvers can overwrite this with
+            a more efficient native implementation
+
+            Arguments:
+                - display: either a list of CPMpy expressions, OR a callback function, called with the variables after value-mapping
+                        default/None: nothing displayed
+                - time_limit: stop after this many seconds (default: None)
+                - solution_limit: stop after this many solutions (default: None)
+                - call_from_model: whether the method is called from a CPMpy Model instance or not
+                - any other keyword argument
+
+            Returns: number of solutions found
+        """
+        if self.objective_given:
+            raise NotImplementedError("Exact does not yet support finding all *optimal* solutions.")
+
+        if not self.solver_is_initialized:
+            self.xct_solver.init([],[])
+            self.solver_is_initialized=True
+            self.xct_solver.setOption("verbosity","0")
+        self.xct_solver.clearAssumptions()
+
+        # set additional keyword arguments
+        for (kw, val) in kwargs.items():
+            self.xct_solver.setOption(kw,str(val))
+
+        solsfound = 0
+        while solution_limit == None or solsfound < solution_limit:
+            # call the solver, with parameters
+            my_status = self.xct_solver.runFull(False,time_limit if time_limit is not None else 0)
+            assert my_status in [0,1,2,3], "Unexpected status code for Exact."
+            if my_status == 0: # found unsatisfiability
+                break
+            elif my_status == 1: # found solution, but not optimality proven
+                assert self.xct_solver.hasSolution()
+                solsfound += 1
+                self._fillObjAndVars()
```

Comment:
these two can go inside the 'if display is not None'

Reply:
Ah, indeed :) Moved `_fillObjAndVars()`, but kept `solsfound+=1`. I can imagine someone wanting to both run a callback and know the total number of solutions.